python/ray/util/ray_executor.py

```diff
@@ -0,0 +1,277 @@
+import itertools
+import time
+from concurrent.futures import Executor
+
+import ray
+from ray.util.annotations import PublicAPI
+
+
+@PublicAPI
+class RayExecutor(Executor):
+    """`RayExecutor` is a drop-in replacement for `ProcessPoolExecutor` and
+    `ThreadPoolExecutor` from `concurrent.futures` but distributes and executes
+    the specified tasks over a Ray cluster instead of multiple processes or
+    threads.
+    """
+
+    _shutdown_lock = False
+    """A dictionary of futures and associated Ray object references generated
+    by the current instance of RayExecutor"""
+    _futures = {}
+
+    def __init__(self, **kwargs):
+        """Initialises a new RayExecutor instance which distributes tasks over
+        a Ray cluster.
+
+        Args:
+            All keyword arguments are passed to ray.init() (see
+            https://docs.ray.io/en/latest/ray-core/package-ref.html#ray-init).
+            For example, this will connect to a cluster at the specified address:
+            .. code-block:: python
+
+                RayExecutor(address='192.168.0.123:25001')
+
+            Note: excluding an address will initialise a local Ray cluster.
+
+        """
+
+        """
+        The following is necessary because `@ray.remote` is only available at runtime.
+        """
+
+        @ray.remote
+        def remote_fn(fn, *args, **kwargs):
+            return fn(*args, **kwargs)
+
+        self.__remote_fn = remote_fn
+        self.context = ray.init(ignore_reinit_error=True, **kwargs)
+
+    @staticmethod
+    def __actor_fn(fn, *args, **kwargs):
+        return fn.remote(*args, **kwargs)
+
+    def submit(self, fn, *args, **kwargs):
+        """Submits a callable to be executed with the given arguments.
+
+        Schedules the callable to be executed as `fn(*args, **kwargs)` and returns
+        a Future instance representing the execution of the callable.
+
+        Returns:
+            A Future representing the given call.
+
+        Usage example:
+
+        .. code-block:: python
+
+            with RayExecutor() as ex:
+                future = ex.submit(lambda x: x * x, 100)
+                result = future.result()
+        """
+        self._check_shutdown_lock()
+        oref = self.__remote_fn.remote(fn, *args, **kwargs)
```

Comment:
Can you 

```
self.__remote_fn.options(name=<function_name>).remote
```
? In this case, the original func name will appear on the dashboard or the state APIs